File: HandOnML/P04_Clustering/03_DBScan/dbscan.py
# K-means clustering

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:, 3:5].values

# Feature Scaling (Optional)
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X = sc_X.fit_transform(X)

# Using dbscan to find frequent data zones
from sklearn.cluster import DBSCAN
dbscan = DBSCAN(eps = 0.2, # Maximum distance between two samples
                min_samples = 5,
                metric = 'euclidean')
dbscan.fit(X)

# Applying dbscan to the mall dataset
y_dbscan = dbscan.fit_predict(X)

# Visualising the clusters
plt.scatter(X[y_dbscan == -1, 0], X[y_dbscan == -1, 1], s = 100, c = 'black', label = 'Non group')
plt.scatter(X[y_dbscan == 1, 0], X[y_dbscan == 1, 1], s = 100, c = 'blue', label = 'Group 1')
plt.scatter(X[y_dbscan == 2, 0], X[y_dbscan == 2, 1], s = 100, c = 'green', label = 'Group 2')
plt.scatter(X[y_dbscan == 3, 0], X[y_dbscan == 3, 1], s = 100, c = 'cyan', label = 'Group 3')
plt.scatter(X[y_dbscan == 4, 0], X[y_dbscan == 4, 1], s = 100, c = 'magenta', label = 'Group 4')
plt.title('Clusters of clients')
plt.xlabel('Annual income')
plt.ylabel('Spending score')
plt.legend()
plt.show()

File: HandOnML/P04_Clustering/03_DBScan/dbscan.R
# DBScan clustering

# Importing the dataset
dataset = read.csv('Mall_Customers.csv')
X = dataset[4:5]

# Using "fpc" library to clustering with dbscan
# install.packages("fpc")
library(fpc)
set.seed(123)
clusters <- fpc::dbscan(X, eps = 5, MinPts = 5)

# Visualising the dbscan clusters with "factoextra"
# install.packages('factoextra')
library(factoextra)
fviz_cluster(clusters, data = X, stand = FALSE,
             ellipse = FALSE, show.clust.cent = FALSE,
             geom = "point",palette = "jco", ggtheme = theme_classic())

# Visualising the clusters with clusterplot
library(cluster)
clusplot(X,
         clusters$cluster,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 2,
         plotchar = FALSE,
         span = TRUE,
         main = paste("Cluster of the clients"),
         xlab = "Annual income",
         ylab = "Spending score")